R/getTimeDivergence.R

```diff
@@ -0,0 +1,98 @@
+#' samples of subject. "n" can be 1 or greater than 1. The time difference
+#' between the n points is also calculated.
+#'
+#' @param se A
+#' \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#' object or
+#' \code{\link[TreeSummarizedExperiment:TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#' @param sample_field column vector containing the all individuals in
+#' `colData` field
+#' @param sample_id character indicating a individual inside of `sample_field`
+#' @param time_field column vector containing the time information
+#' in `colData` field
+#' @param time_interval integer value indicating the increment between the time
+#' steps. It can be 1 or higher than 1.
+#' @param transposed logical scalar assigning samples to rows if they are in
+#' columns. If samples are not in rows, the default value is applied.
+#' (default: \code{transposed = FALSE})
+#'
+#' @return a matrix containing the beta diversity and time difference
+#' between samples in n time steps
+#'
+#' @importFrom mia calculateDistance
+#'
+#' @examples
+#' library(microbiomeDataSets)
+#' tse <- GrieneisenTSData()
+#'
+#' BaboonDivergence <- getTimeDivergence(tse, sample_field = "baboon_id",
+#'                                     sample_id = "Baboon_1" ,
+#'                                     time_field = "collection_date",
+#'                                     time_interval = 1,
+#'                                     transposed = FALSE )
+#'
+#'@export
+getTimeDivergence <- function(se, sample_field, sample_id, time_field, time_interval, transposed = FALSE ){
+
+    df <-as.data.frame(colData(se)[, c(sample_field, time_field)])
+
+    #per sample
+    sample_df <- df[which((df[, sample_field] == sample_id) == TRUE),]
+
+    mat <- data.matrix(sample_df)
+
+    #time in increasing order
+    mat[, time_field] <- mat[, time_field][order(mat[, time_field], decreasing = FALSE)]
+
+     tt <- sapply(seq_len(nrow(mat)), FUN = function(i){
+      k <- c(i, i + time_interval)
+      i <- i + 1
+      k
+     })
+
+    #total number of sample pairs
+    total <- nrow(mat) - time_interval
+
+    location <- tt[,seq_len(total)]
+
+    #location matrix turned into list object
+    if(length(location) > 2){
+        list <- lapply(seq_len(ncol(location)), function(i) location[,i])
+
+        samplename <-lapply(seq_len(length(list)), function(i){ mat[, sample_field][list[[i]]]})
+
+        time <-lapply(seq_len(length(list)), function(i){ mat[, time_field][list[[i]]]})
+
+        #related samples are chosen from assay
+        list_t <- lapply(seq_len(length(time)), function(i){ as.matrix(assay(se)[,names(samplename[[i]])])})
+
+        if(!transposed){
+          list_t <- lapply(seq_len(length(list_t)), function(i){
+            t(list_t[[i]])
+          })
+        }
+
+        timedivergence_n <- vapply(seq_len(length(list_t)),
+                                   FUN = function(i) {calculateDistance(list_t[[i]])},
```

Comment:
The latest version of *mia* says

Warning message:
'calculateDistance' is deprecated. 
Instead, use directly the function that is specified by 'FUN' argument. 

The FUN argument refers to `stats::dist` 

-> Let us use that one.

Then, with the time interval t, should this be just simple as:

```
t <- 1 
mat <- matrix(1:60, nrow = 5) 
timedivergence_n <- sapply((t+1):nrow(mat), function (i) {stats::dist(mat[c(i, i-t), ])})
```

Reply:
(or something like that for `vapply`)

Reply:
I have been trying to implement the method which splits SE by chosen field, applies the operation, and puts it back. But there is an error that I couldn't fix for a while. Here I used `time` as the field to split SE object, but it can be anything as you suggested. 

```
data(hitchip1006)
se <- hitchip1006
field <- "time"

spl <- split(colnames(se), colData(se)[, field])
  
se_list <- lapply(seq_len(length(spl)), function (i) {SummarizedExperiment(assays = as.matrix(assay(se)[,spl[[i]]]),
                                                                             colData = colData(se)[spl[[i]],])})
   
mat <- sapply(seq_len(length(se_list)), function(i){t(assay(se_list[[i]]))})
   
t <- 1

sapply(seq_len(length(mat)), function(n){ 
        sapply((t+1):nrow(mat[[n]]), function (i) {stats::dist(mat[[n]][c(i, i-t), ])})})
```
error is in `sapply` part, I tried with for loop, and it gives the same error. `Error in mat[[n]][c(i, i - t), ] : subscript out of bounds`
 ` i ` stays in 2L and `s` stays in 6L, and it is because mat[[6]] has only one sample. Then I tried something like this 

`sapply(seq_len(length(mat)), function(n){ 
      if( nrow(mat[[n]]) > 1 ){
          sapply((t+1):nrow(mat[[i]]), function (i) {stats::dist(mat[[i]][c(i, i-t), ])}
          )
      } else if ( nrow(mat[[n]]) == 1 ) { 
        next
        }
  }
)`

That also didn't work. I am out of solutions currently. Anyhow, I can do the same operation without splitting the SE but instead making data frame from the chosen field.

Reply:
Ok how about this one:

Reply:
```
library(miaTime)
data(hitchip1006)
se <- hitchip1006
field <- "subject"


# Define the time step of interest
step <- 1
myassay <- "counts"

# Define the names for the newly added fields
# for the time difference and community dissimilarity
new_field <- "time_divergence"
new_field2 <- "time_difference"

# Dissimilarity will be calculated between rows, as in stats::dist
check_pairwise_dist <- function (x, distfun, step, myassay, new_field) {

    mat <- t(assay(x, myassay))

    # Add new field to coldata
    colData(x)[, new_field] <- rep(NA, nrow(mat))
    
    if (nrow(mat) > step) {
         inds <- (step+1):nrow(mat)
         d <- sapply(inds, function (i) {distfun(mat[c(i, i-step), ])})
	 colData(x)[inds, new_field] <- d
    }
    return(x)
}

# Split SE into a list, by subject
spl <- split(colnames(se), colData(se)[, field])
se_list <- lapply(seq_len(length(spl)), function (i) {se[, spl[[i]]]})

# Manipulate each subobject
se_list <- lapply(seq_along(se_list),
    function(i){ check_pairwise_dist(se_list[[i]], distfun=stats::dist, step, myassay, new_field) })

# Merge the objects back into a single SE
se <- do.call("cbind", se_list)
```

Reply:
It is a bit slow, but acceptable for a first version. Time difference to be added, and material to be polished.

Reply:
Thanks! I'll complete what is missing and look for a faster implementation.

Reply:
I have just realized that it gives the same error. Because again, `x` is only one subject, so the distance cannot be applied in this way.

Reply:
```
inds <- (time_interval+1):length(se_list)
      d <- sapply(inds, function (i) { s <- t(assay(se_list[[i]]))
                                       ss <- t(assay(selist[[i-time_interval]]))
                                       dd <- rbind(s, ss)
                                       distfun(dd))})
```
But I found a different way which does the job. I will add everything soon.